flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala

```diff
@@ -0,0 +1,206 @@
+      new ValueStateDescriptor[Long]("dataCountState", classOf[Long])
+    dataCountState = getRuntimeContext.getState(dataCountStateDescriptor)
+
+    val accumulatorStateDescriptor =
+      new ValueStateDescriptor[Row]("accumulatorState", aggregationStateType)
+    accumulatorState = getRuntimeContext.getState(accumulatorStateDescriptor)
+
+    val mapStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]](
+        "dataState",
+        classOf[Long],
+        classOf[JList[Row]])
+
+    dataState = getRuntimeContext.getMapState(mapStateDescriptor)
+
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+
+    // triggering timestamp for trigger calculation
+    val triggeringTs = ctx.timestamp
+
+    val lastTriggeringTs = lastTriggeringTsState.value
+    // check if the data is expired, if not, save the data and register event time timer
+    if (triggeringTs > lastTriggeringTs && triggeringTs > ctx.timerService.currentWatermark) {
+      if (dataState.contains(triggeringTs)) {
+        val data = dataState.get(triggeringTs)
+        data.add(input)
+        dataState.put(triggeringTs, data)
+      } else {
+        val data = new ArrayList[Row]
+        data.add(input)
+        dataState.put(triggeringTs, data)
+        // register event time timer
+        ctx.timerService.registerEventTimeTimer(triggeringTs)
+      }
+    }
+  }
+
+  override def onTimer(
+    timestamp: Long,
+    ctx: ProcessFunction[Row, Row]#OnTimerContext,
+    out: Collector[Row]): Unit = {
+
+    // gets all window data from state for the calculation
+    val inputs: JList[Row] = dataState.get(timestamp)
+    if (null != inputs) {
+      var accumulators = accumulatorState.value
+      var j: Int = 0
+      while (j < inputs.size) {
+        val input = inputs.get(j)
+
+        // initialize when first run or failover recovery per key
+        if (null == accumulators) {
+          accumulators = new Row(aggregates.length)
+          var i = 0
+          while (i < aggregates.length) {
+            accumulators.setField(i, aggregates(i).createAccumulator)
+            i += 1
+          }
+        }
+
+        var dataCount = dataCountState.value + 1
+        dataCountState.update(dataCount)
+
+        var lastExpiredRow: Row = null
+
+        if (dataCount > precedingOffset) {
+          val dataTimestampIt = dataState.keys.iterator
+          var expiredDataTs: Long = Long.MaxValue
+          while (dataTimestampIt.hasNext) {
+            val dataTs = dataTimestampIt.next
+            if (dataTs < expiredDataTs) {
+              expiredDataTs = dataTs
+            }
+          }
+          val windowDataList = dataState.get(expiredDataTs)
+
+          lastExpiredRow = windowDataList.get(0)
+
+          windowDataList.remove(0)
+
+          if (windowDataList.size > 0) {
+            dataState.put(expiredDataTs, windowDataList)
```

Comment:
We should only update the state (MapState.put) at the end of the method or remove if the `retractList` became empty.